docs/ActiveLabel.rst

```diff
@@ -0,0 +1,135 @@
+ActiveLabel
+===========
+
+``ActiveNode`` provides an ability to define inheritance of models which also gives subclasess the labels of their parent models.  In Ruby, however, inheritence of classes is not sufficient.  Sometimes is makes more sense to be able to build a module which defines behavior (or "concerns") which could be applied to any model.  This is what ``ActiveLabel`` provides.
```

Comment:
I'd replace with:
The Neo4jrb gem provides several ways to keep your code DRY. Subclasses can inherit from an ``ActiveNode`` model, which gives subclasses the labels of their parent models in addition to methods from the parent models. However, inheritance alone is not always sufficient.  Sometimes what you want to do is conditionally add a module to an ``ActiveNode`` model if a specific label is present on an instance of that model. For an example of when this is needed, look at the Neo4j's example movie database (https://neo4j.com/developer/movie-database/). ``ActiveLabel`` allows you to accomplish this.

Reply:
I still don't understand the conditional part.

Suppose you have a person model who may be an Actor or a Director or both.

1. You still have to be careful not to have namespace clash.
Regardless of having the label or not, you still need associations like directed_movies, acted_movies for each relationship type. If each module called the association `movies`...

2. Is there any benefit to hiding these chunks of code when the label is not set??
Class methods for associations are called when loading the models, right? and then instances are made. Why bother conditionally adding in the the methods? Won't you have to pepper your code with checks if the methods exist and if the label is set anyway?

Reply:
Regarding your first point, you'd only need a `directed_movies` association if someone was a director (i.e. has the `:Director` label). Put another way, if `person.director? == false` then `person.directed_movies` would return an unknown method error.

Regarding your second point, if you don't see a benefit to conditionally adding code based on the presence of a label, what have you been arguing for this whole time? Honestly curious. It seems clear that we've been operating with different goals.

Reply:
Perhaps I need to go back and re-read some of your comments, given your second point above...I've missed something.

Reply:
You're absolutely correct. I never really caught the conditional argument. My bad.

But my concern is, as a developer one has to check `person.director?` before called these methods or one has to handle the unknown method error. What I'm trying to understand then is why conditionally include the code if making the checks anyway?

Reply:
Oh ya, you absolutely need to check for `person.director?` Or in my case, I'd need to check for `person.current?` or `person.destroyed?` and handle these two cases differently. I always need to do this though. A current person node is conceptually different from a destroyed person node and needs to be handled different. Its appropriate that different methods exist on each. 

It's conceptually similar to checking what type of object a variable contains.

As far as why I'd want to conditionally include the code, well mainly because a `current` node might have a method of the same name as a `destroyed` node, but the two methods do completely different things. I'd want to check to see what type of node I was dealing with, so I knew what the method did. Obviously you could achieve similar results with some type of method overloading, but there are always multiple ways to accomplish something in ~~ruby~~ programming.

Reply:
Also worth noting: if your query only returns `:Person:Actor` nodes, then you _know_ that the resulting objects have the `:Actor` methods.

Reply:
Ok, but in the case of :Actor and :Director, a :Person could be both, and now if each one has a method with the same name, you have a clash.

Reply:
True, but I think that's a limitation of Ruby modules, rather than a flaw with this implementation strategy. If two modules define a method with the same name, and you include both of them in a class, the last one will overwrite the others.

Reply:
Put another way, if a method might apply to either an :Actor or :Director, you need to be more thoughtful about the naming than if it just applies to one or the other.